app/models/spree/payment/processing.rb

```diff
@@ -0,0 +1,242 @@
+      end
+
+      def void_transaction!
+        return true if void?
+        protect_from_connection_error do
+          check_environment
+
+          if payment_method.payment_profiles_supported?
+            # Gateways supporting payment profiles will need access to credit card object because this stores the payment profile information
+            # so supply the authorization itself as well as the credit card, rather than just the authorization code
+            response = payment_method.void(self.response_code, source, gateway_options)
+          else
+            # Standard ActiveMerchant void usage
+            response = payment_method.void(self.response_code, gateway_options)
+          end
+          record_response(response)
+
+          if response.success?
+            self.response_code = response.authorization
+            self.void
+          else
+            gateway_error(response)
+          end
+        end
+      end
+
+      def credit!(credit_amount=nil)
+        protect_from_connection_error do
+          check_environment
+
+          credit_amount ||= credit_allowed >= order.outstanding_balance.abs ? order.outstanding_balance.abs : credit_allowed.abs
+          credit_amount = credit_amount.to_f
+
+          if payment_method.payment_profiles_supported?
+            response = payment_method.credit((credit_amount * 100).round, source, response_code, gateway_options)
+          else
+            response = payment_method.credit((credit_amount * 100).round, response_code, gateway_options)
+          end
+
+          record_response(response)
+
+          if response.success?
+            self.class.create(
+              :order => order,
+              :source => self,
+              :payment_method => payment_method,
+              :amount => credit_amount.abs * -1,
+              :response_code => response.authorization,
+              :state => 'completed'
+            )
+          else
+            gateway_error(response)
+          end
+        end
+      end
+
+      def refund!(refund_amount = nil)
+        protect_from_connection_error do
+          check_environment
+
+          refund_amount = calculate_refund_amount(refund_amount)
+
+          if payment_method.payment_profiles_supported?
+            response = payment_method.refund((refund_amount * 100).round, source, response_code, gateway_options)
+          else
+            response = payment_method.refund((refund_amount * 100).round, response_code, gateway_options)
+          end
+
+          record_response(response)
+
+          if response.success?
+            self.class.create(order: order,
+                              source: self,
+                              payment_method: payment_method,
+                              amount: refund_amount.abs * -1,
+                              response_code: response.authorization,
+                              state: 'completed')
+          else
+            gateway_error(response)
+          end
```

Comment:
this code above is copy paste from credit! above, so many things to improve. it even feels nice :blush:

Reply:
I leave that out this PR because things will become unreviewable and IMO this should be better solved extracting services out of it.